File: examples/Python/prediction_intervals.py
"""Prediction intervals example.

Run with:
    pip install linreg-core
    python prediction_intervals.py

Demonstrates:
- OLS prediction intervals (exact, using leverage)
- Ridge / Lasso / Elastic Net prediction intervals (conservative approximation)
- How interval width varies with confidence level
- How intervals widen for out-of-sample (extrapolation) points
"""

import linreg_core


def main():
    print("╔══════════════════════════════════════════════════════════════════════╗")
    print("║                    PREDICTION INTERVALS                              ║")
    print("╚══════════════════════════════════════════════════════════════════════╝")
    print()

    # Training data: house size (sqft/100) vs price ($k)
    sqft  = [10.0, 12.0, 14.0, 16.0, 18.0, 20.0, 22.0, 24.0, 26.0, 28.0]
    price = [150.0, 175.0, 210.0, 240.0, 265.0, 295.0, 330.0, 360.0, 400.0, 430.0]

    # New points: in-sample range + one extrapolation at sqft=35
    new_sqft_vals = [11.0, 15.0, 19.0, 25.0, 35.0]
    new_x = [new_sqft_vals]   # list of column lists, matching x_vars format

    print("Training: 10 houses, sqft (hundreds) -> price ($k)")
    print(f"Predicting at: {new_sqft_vals}")
    print("Note: sqft=35 is extrapolation (training range: 10–28)")
    print()

    # ── 1. OLS Prediction Intervals ───────────────────────────────────────────
    print("━━━ 1. OLS Prediction Intervals (95%, exact) ━━━━━━━━━━━━━━━━━━━━━━━━")
    print("  Formula: PI = ŷ ± t(α/2, df) × √(MSE × (1 + leverage))")
    print()

    ols_pi = linreg_core.ols_prediction_intervals(price, [sqft], new_x, alpha=0.05)

    print(f"  {'SqFt':>8}  {'Predicted':>10}  {'Lower 95%':>10}  "
          f"{'Upper 95%':>10}  {'Width':>10}  {'Leverage':>8}")
    print(f"  {'─'*64}")
    for i, sx in enumerate(new_sqft_vals):
        width  = ols_pi.upper_bound[i] - ols_pi.lower_bound[i]
        marker = " <-- extrapolation" if sx > 28.0 else ""
        print(f"  {sx:>8.1f}  {ols_pi.predicted[i]:>10.2f}  "
              f"{ols_pi.lower_bound[i]:>10.2f}  {ols_pi.upper_bound[i]:>10.2f}  "
              f"{width:>10.2f}  {ols_pi.leverage[i]:>8.4f}{marker}")
    print()
    print("  Note: Interval width grows at the extrapolation point (sqft=35)")
    print("        because leverage is high far from the training data centre.")
    print(f"  df_residuals: {ols_pi.df_residuals:.1f}")
    print()

    # ── 2. Interval Width vs Confidence Level ─────────────────────────────────
    print("━━━ 2. Interval Width vs Confidence Level (at sqft=19) ━━━━━━━━━━━━━━")
    print(f"  {'Confidence':>12}  {'Lower':>10}  {'Predicted':>10}  "
          f"{'Upper':>10}  {'Width':>10}")
    print(f"  {'─'*56}")
    for conf in [0.50, 0.80, 0.90, 0.95, 0.99]:
        pi = linreg_core.ols_prediction_intervals(
            price, [sqft], [[19.0]], alpha=1.0 - conf
        )
        width = pi.upper_bound[0] - pi.lower_bound[0]
        print(f"  {conf*100:>11.0f}%  {pi.lower_bound[0]:>10.2f}  "
              f"{pi.predicted[0]:>10.2f}  {pi.upper_bound[0]:>10.2f}  {width:>10.2f}")
    print()

    # ── 3. Regularized Model Prediction Intervals ─────────────────────────────
    print("━━━ 3. Regularized Model Prediction Intervals (at sqft=19) ━━━━━━━━━━")
    print("  (Conservative approximation: unpenalized leverage + penalized MSE)")
    print()

    new_x_single = [[19.0]]

    ridge_pi = linreg_core.ridge_prediction_intervals(
        price, [sqft], new_x_single, alpha=0.05, lambda_val=1.0, standardize=True
    )
    lasso_pi = linreg_core.lasso_prediction_intervals(
        price, [sqft], new_x_single, alpha=0.05, lambda_val=0.5, standardize=True
    )
    enet_pi = linreg_core.elastic_net_prediction_intervals(
        price, [sqft], new_x_single, alpha=0.05,
        lambda_val=0.5, enet_alpha=0.5, standardize=True
    )

    print(f"  {'Method':<14}  {'Predicted':>10}  {'Lower 95%':>10}  "
          f"{'Upper 95%':>10}  {'Width':>10}")
    print(f"  {'─'*58}")

    # OLS at sqft=19 (index 2 in the original new_x prediction)
    ols_width = ols_pi.upper_bound[2] - ols_pi.lower_bound[2]
    print(f"  {'OLS (exact)':<14}  {ols_pi.predicted[2]:>10.2f}  "
          f"{ols_pi.lower_bound[2]:>10.2f}  {ols_pi.upper_bound[2]:>10.2f}  {ols_width:>10.2f}")

    r_width = ridge_pi.upper_bound[0] - ridge_pi.lower_bound[0]
    print(f"  {'Ridge':<14}  {ridge_pi.predicted[0]:>10.2f}  "
          f"{ridge_pi.lower_bound[0]:>10.2f}  {ridge_pi.upper_bound[0]:>10.2f}  {r_width:>10.2f}")

    l_width = lasso_pi.upper_bound[0] - lasso_pi.lower_bound[0]
    print(f"  {'Lasso':<14}  {lasso_pi.predicted[0]:>10.2f}  "
          f"{lasso_pi.lower_bound[0]:>10.2f}  {lasso_pi.upper_bound[0]:>10.2f}  {l_width:>10.2f}")

    e_width = enet_pi.upper_bound[0] - enet_pi.lower_bound[0]
    print(f"  {'Elastic Net':<14}  {enet_pi.predicted[0]:>10.2f}  "
          f"{enet_pi.lower_bound[0]:>10.2f}  {enet_pi.upper_bound[0]:>10.2f}  {e_width:>10.2f}")

    print()
    print("  Note: Regularized intervals are conservative (wider) because they")
    print("        use unpenalized leverage with the penalized model's MSE.")
    print()

    # ── 4. Standard Errors ────────────────────────────────────────────────────
    print("━━━ 4. OLS Prediction Standard Errors ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━")
    print(f"  {'SqFt':>8}  {'Predicted':>10}  {'SE(pred)':>10}  {'Leverage':>10}")
    print(f"  {'─'*44}")
    for i, sx in enumerate(new_sqft_vals):
        print(f"  {sx:>8.1f}  {ols_pi.predicted[i]:>10.2f}  "
              f"{ols_pi.se_pred[i]:>10.4f}  {ols_pi.leverage[i]:>10.4f}")


if __name__ == "__main__":
    main()

File: examples/Rust/prediction_intervals.rs
//! Prediction intervals example.
//!
//! Run with:
//!   cargo run --example prediction_intervals --no-default-features
//!
//! Demonstrates:
//! - OLS prediction intervals (exact, using leverage)
//! - Ridge / Lasso / Elastic Net prediction intervals (conservative approximation)
//! - Difference between prediction intervals (new obs) vs confidence intervals (mean)
//! - How intervals widen for out-of-sample (extrapolation) points

use linreg_core::linalg::Matrix;
use linreg_core::prediction_intervals::{
    elastic_net_prediction_intervals, lasso_prediction_intervals, prediction_intervals,
    ridge_prediction_intervals,
};
use linreg_core::regularized::{
    elastic_net_fit, lasso_fit, ridge_fit, ElasticNetOptions, LassoFitOptions, RidgeFitOptions,
};

fn main() {
    // ── Training data: house size (sqft/100) vs price ($k) ─────────────────────
    let sqft  = vec![10.0, 12.0, 14.0, 16.0, 18.0, 20.0, 22.0, 24.0, 26.0, 28.0];
    let price = vec![150.0, 175.0, 210.0, 240.0, 265.0, 295.0, 330.0, 360.0, 400.0, 430.0];

    // New points to predict: in-sample range and one extrapolation
    let new_sqft_vals = vec![11.0, 15.0, 19.0, 25.0, 35.0]; // 35 is extrapolation
    let new_sqft: Vec<&[f64]> = vec![&new_sqft_vals];

    println!("╔══════════════════════════════════════════════════════════════════════╗");
    println!("║                    PREDICTION INTERVALS                              ║");
    println!("╚══════════════════════════════════════════════════════════════════════╝");
    println!();
    println!("Training: 10 houses, sqft (hundreds) -> price ($k)");
    println!("Predicting at: {:?}", new_sqft_vals);
    println!("Note: sqft=35 is extrapolation (training range: 10-28)");
    println!();

    // ── 1. OLS Prediction Intervals ────────────────────────────────────────────
    println!("━━━ 1. OLS Prediction Intervals (95%, exact) ━━━━━━━━━━━━━━━━━━━━━━━━");
    println!("  Formula: PI = ŷ ± t(α/2, df) × √(MSE × (1 + leverage))");
    println!();

    let ols_pi = prediction_intervals(&price, &[sqft.clone()], &new_sqft, 0.05)
        .expect("OLS prediction intervals failed");

    println!("  {:>8}  {:>10}  {:>10}  {:>10}  {:>10}  {:>8}",
             "SqFt", "Predicted", "Lower 95%", "Upper 95%", "Width", "Leverage");
    println!("  {}", "─".repeat(64));
    for i in 0..new_sqft_vals.len() {
        let width = ols_pi.upper_bound[i] - ols_pi.lower_bound[i];
        let extrap = if new_sqft_vals[i] > 28.0 { " <-- extrapolation" } else { "" };
        println!(
            "  {:>8.1}  {:>10.2}  {:>10.2}  {:>10.2}  {:>10.2}  {:>8.4}{}",
            new_sqft_vals[i],
            ols_pi.predicted[i],
            ols_pi.lower_bound[i],
            ols_pi.upper_bound[i],
            width,
            ols_pi.leverage[i],
            extrap
        );
    }
    println!();
    println!("  Note: Interval width grows at the extrapolation point (sqft=35)");
    println!("        because leverage is high far from the training data center.");
    println!("  df_residuals: {}", ols_pi.df_residuals);
    println!();

    // ── 2. Effect of alpha (confidence level) ─────────────────────────────────
    println!("━━━ 2. Interval Width vs Confidence Level (at sqft=19) ━━━━━━━━━━━━━━");
    println!("  {:>12}  {:>10}  {:>10}  {:>10}  {:>10}",
             "Confidence", "Lower", "Predicted", "Upper", "Width");
    println!("  {}", "─".repeat(56));
    for &conf in &[0.50, 0.80, 0.90, 0.95, 0.99] {
        let alpha = 1.0 - conf;
        let pi = prediction_intervals(&price, &[sqft.clone()], &[&[19.0_f64][..]], alpha)
            .expect("PI failed");
        let width = pi.upper_bound[0] - pi.lower_bound[0];
        println!(
            "  {:>11.0}%  {:>10.2}  {:>10.2}  {:>10.2}  {:>10.2}",
            conf * 100.0, pi.lower_bound[0], pi.predicted[0], pi.upper_bound[0], width
        );
    }
    println!();

    // ── 3. Regularized model prediction intervals ──────────────────────────────
    println!("━━━ 3. Regularized Model Prediction Intervals (at sqft=19) ━━━━━━━━━━");
    println!("  (Conservative approximation using unpenalized leverage + fit MSE)");
    println!();

    let mut x_data = Vec::with_capacity(10 * 2);
    for i in 0..10 {
        x_data.push(1.0);
        x_data.push(sqft[i]);
    }
    let x_mat = Matrix::new(10, 2, x_data);
    let x_vars = vec![sqft.clone()];
    let new_x_single: Vec<&[f64]> = vec![&new_sqft_vals[2..3]]; // sqft=19

    // Ridge
    let ridge = ridge_fit(&x_mat, &price, &RidgeFitOptions {
        lambda: 1.0, standardize: true, intercept: true, ..Default::default()
    }).expect("Ridge failed");
    let ridge_pi = ridge_prediction_intervals(&ridge, &x_vars, &new_x_single, 0.05)
        .expect("Ridge PI failed");

    // Lasso
    let lasso = lasso_fit(&x_mat, &price, &LassoFitOptions {
        lambda: 0.5, standardize: true, intercept: true, ..Default::default()
    }).expect("Lasso failed");
    let lasso_pi = lasso_prediction_intervals(&lasso, &x_vars, &new_x_single, 0.05)
        .expect("Lasso PI failed");

    // Elastic Net
    let enet = elastic_net_fit(&x_mat, &price, &ElasticNetOptions {
        lambda: 0.5, alpha: 0.5, standardize: true, intercept: true, ..Default::default()
    }).expect("Elastic Net failed");
    let enet_pi = elastic_net_prediction_intervals(&enet, &x_vars, &new_x_single, 0.05)
        .expect("Elastic Net PI failed");

    println!("  {:14}  {:>10}  {:>10}  {:>10}  {:>10}",
             "Method", "Predicted", "Lower 95%", "Upper 95%", "Width");
    println!("  {}", "─".repeat(58));

    let ols_at_19 = &ols_pi; // already computed at index 2 (sqft=19)
    let ols_width = ols_at_19.upper_bound[2] - ols_at_19.lower_bound[2];
    println!("  {:14}  {:>10.2}  {:>10.2}  {:>10.2}  {:>10.2}",
             "OLS (exact)",
             ols_at_19.predicted[2], ols_at_19.lower_bound[2],
             ols_at_19.upper_bound[2], ols_width);

    let r_width = ridge_pi.upper_bound[0] - ridge_pi.lower_bound[0];
    println!("  {:14}  {:>10.2}  {:>10.2}  {:>10.2}  {:>10.2}",
             "Ridge",
             ridge_pi.predicted[0], ridge_pi.lower_bound[0],
             ridge_pi.upper_bound[0], r_width);

    let l_width = lasso_pi.upper_bound[0] - lasso_pi.lower_bound[0];
    println!("  {:14}  {:>10.2}  {:>10.2}  {:>10.2}  {:>10.2}",
             "Lasso",
             lasso_pi.predicted[0], lasso_pi.lower_bound[0],
             lasso_pi.upper_bound[0], l_width);

    let e_width = enet_pi.upper_bound[0] - enet_pi.lower_bound[0];
    println!("  {:14}  {:>10.2}  {:>10.2}  {:>10.2}  {:>10.2}",
             "Elastic Net",
             enet_pi.predicted[0], enet_pi.lower_bound[0],
             enet_pi.upper_bound[0], e_width);

    println!();
    println!("  Note: Regularized intervals are conservative (wider) because they");
    println!("        use unpenalized leverage with the penalized model's MSE.");
}
